### PR TITLE
TVDB: Fetch image keytypes before querying images

### PR DESCRIPTION
**Pull request by cvium**
**Changes**
Use the `/series/{id}/images` endpoint to get the available image types eg. poster, banners to avoid making a lot of unnecessary requests that result in exceptions being thrown.
Also fix the bad banner url in the series provider.

### CODE REVIEW

File: MediaBrowser.Providers/Plugins/TheTvdb/TvdbClientManager.cs [hunk 1]
```diff
@@ -229,6 +229,51 @@ await GetEpisodesPageAsync(Convert.ToInt32(seriesTvdbId), episodeQuery, language
             return GetEpisodesPageAsync(tvdbId, 1, episodeQuery, language, cancellationToken);
         }
 
+        public async Task<IEnumerable<KeyType>> GetImageKeyTypesForSeriesAsync(int tvdbId, string language, CancellationToken cancellationToken)
+        {
+            var cacheKey = GenerateKey(nameof(TvDbClient.Series.GetImagesSummaryAsync), tvdbId);
+            var imagesSummary = await TryGetValue(cacheKey, language, () => TvDbClient.Series.GetImagesSummaryAsync(tvdbId, cancellationToken)).ConfigureAwait(false);
+            var keyTypes = new List<KeyType>();
```
Comment (crobibero) [inline MediaBrowser.Providers/Plugins/TheTvdb/TvdbClientManager.cs:236]:
```suggestion
```

File: MediaBrowser.Providers/Plugins/TheTvdb/TvdbClientManager.cs [hunk 2]
```diff
@@ -229,6 +229,51 @@ await GetEpisodesPageAsync(Convert.ToInt32(seriesTvdbId), episodeQuery, language
...
+
+            if (imagesSummary.Data.Fanart > 0)
+            {
+                keyTypes.Add(KeyType.Fanart);
```
Comment (crobibero) [inline MediaBrowser.Providers/Plugins/TheTvdb/TvdbClientManager.cs:240]:
```suggestion
                yield return KeyType.Fanart;
```

File: MediaBrowser.Providers/Plugins/TheTvdb/TvdbClientManager.cs [hunk 3]
```diff
@@ -229,6 +229,51 @@ await GetEpisodesPageAsync(Convert.ToInt32(seriesTvdbId), episodeQuery, language
...
+            }
+
+            if (imagesSummary.Data.Series > 0)
+            {
+                keyTypes.Add(KeyType.Series);
```
Comment (crobibero) [inline MediaBrowser.Providers/Plugins/TheTvdb/TvdbClientManager.cs:245]:
```suggestion
                yield return KeyType.Series;
```

File: MediaBrowser.Providers/Plugins/TheTvdb/TvdbClientManager.cs [hunk 4]
```diff
@@ -229,6 +229,51 @@ await GetEpisodesPageAsync(Convert.ToInt32(seriesTvdbId), episodeQuery, language
...
+            }
+
+            if (imagesSummary.Data.Poster > 0)
+            {
+                keyTypes.Add(KeyType.Poster);
```
Comment (crobibero) [inline MediaBrowser.Providers/Plugins/TheTvdb/TvdbClientManager.cs:250]:
```suggestion
                yield return KeyType.Poster;
```

File: MediaBrowser.Providers/Plugins/TheTvdb/TvdbClientManager.cs [hunk 5]
```diff
@@ -229,6 +229,51 @@ await GetEpisodesPageAsync(Convert.ToInt32(seriesTvdbId), episodeQuery, language
...
+            }
+
+            return keyTypes;
```
Comment (crobibero) [inline MediaBrowser.Providers/Plugins/TheTvdb/TvdbClientManager.cs:253]:
```suggestion
```

File: MediaBrowser.Providers/Plugins/TheTvdb/TvdbClientManager.cs [hunk 6]
```diff
@@ -229,6 +229,51 @@ await GetEpisodesPageAsync(Convert.ToInt32(seriesTvdbId), episodeQuery, language
...
+        }
+
+        public async Task<IEnumerable<KeyType>> GetImageKeyTypesForSeasonAsync(int tvdbId, string language, CancellationToken cancellationToken)
+        {
+            var cacheKey = GenerateKey(nameof(TvDbClient.Series.GetImagesSummaryAsync), tvdbId);
+            var imagesSummary = await TryGetValue(cacheKey, language, () => TvDbClient.Series.GetImagesSummaryAsync(tvdbId, cancellationToken)).ConfigureAwait(false);
+            var keyTypes = new List<KeyType>();
```
Comment (crobibero) [inline MediaBrowser.Providers/Plugins/TheTvdb/TvdbClientManager.cs:260]:
```suggestion
```

File: MediaBrowser.Providers/Plugins/TheTvdb/TvdbClientManager.cs [hunk 7]
```diff
@@ -229,6 +229,51 @@ await GetEpisodesPageAsync(Convert.ToInt32(seriesTvdbId), episodeQuery, language
...
+
+            if (imagesSummary.Data.Season > 0)
+            {
+                keyTypes.Add(KeyType.Season);
```
Comment (crobibero) [inline MediaBrowser.Providers/Plugins/TheTvdb/TvdbClientManager.cs:264]:
```suggestion
                yield return KeyType.Season;
```

File: MediaBrowser.Providers/Plugins/TheTvdb/TvdbClientManager.cs [hunk 8]
```diff
@@ -229,6 +229,51 @@ await GetEpisodesPageAsync(Convert.ToInt32(seriesTvdbId), episodeQuery, language
...
+            }
+
+            if (imagesSummary.Data.Fanart > 0)
+            {
+                keyTypes.Add(KeyType.Fanart);
```
Comment (crobibero) [inline MediaBrowser.Providers/Plugins/TheTvdb/TvdbClientManager.cs:269]:
```suggestion
                yield return KeyType.Fanart;
```

File: MediaBrowser.Providers/Plugins/TheTvdb/TvdbClientManager.cs [hunk 9]
```diff
@@ -229,6 +229,51 @@ await GetEpisodesPageAsync(Convert.ToInt32(seriesTvdbId), episodeQuery, language
...
+            }
+
+            // TODO seasonwide is not supported in TvDbSharper
+
+            return keyTypes;
```
Comment (crobibero) [inline MediaBrowser.Providers/Plugins/TheTvdb/TvdbClientManager.cs:274]:
```suggestion
```

File: MediaBrowser.Providers/Plugins/TheTvdb/TvdbClientManager.cs (hunk 2 above)
Comment (cvium) [reply MediaBrowser.Providers/Plugins/TheTvdb/TvdbClientManager.cs:240]: Just fyi, this won't work with a Task return type. Doesn't matter though. I updated this way before your review comments, but github refuses to update the PR.

**Pull request merged**